share/sysbuild/CMakeLists.txt

```diff
@@ -34,11 +34,42 @@ sysbuild_add_subdirectory(images)
 get_property(IMAGES GLOBAL PROPERTY sysbuild_images)
 sysbuild_module_call(PRE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES})
 sysbuild_images_order(IMAGES_CONFIGURATION_ORDER CONFIGURE IMAGES ${IMAGES})
+
+# This file is generated piece by piece. It will contain information about the
+# current state of this multi-image build, for the purpose of integration with
+# west and other utilities.
+set(DOMAINS_YAML ${CMAKE_BINARY_DIR}/domains.yaml)
+file(WRITE ${DOMAINS_YAML}
```

Comment:
> > in case it ever gets interrupted.
>
> then the sysbuild CMake configure has failed, which again means that no build file (build.ninja/Makefile) has been generated.

But there could be. If an image has finished configuring, it will have its own `build.ninja` file.

> Having a partially created domains.yaml could even lead west to believe there is a valid build system, which is not the case.

But there could be. Remember that with sysbuild, you aren't dealing with just one build system, but several. `domains.yaml` is supposed to tell you about all of them, and west is still able to verify that each one is a valid Zephyr build.

> In which way is this useful, compared to the maintenance of sysbuild itself ?

This is hard for me to argue, which is why I've included the "build pristine" patch to help motivate this (and there can be other use cases as well).

The main idea is that in sysbuild, you're naturally dealing with multiple build systems - one for each image, and one for sysbuild itself. All of these are generated at different times. In particular, each image has to finish configuration before sysbuild can do so. But, if sysbuild itself doesn't finish, you can still have a couple of build systems in your directory, which are practically self-contained and ready to go.

There can be situations (and I've identified one of them) where you might want to know about these individual build systems, without this being tied to whether sysbuild has completed its job.

> Imo it makes the top-level CMake file harder to read

I've considered hiding the YAML code behind a few functions, but I wasn't sure where to put them. I'm open to improving readability that way.

Reply:
> > Having a partially created domains.yaml could even lead west to believe there is a valid build system, which is not the case.
> 
> But there could be. Remember that with sysbuild, you aren't dealing with just one build system, but several. `domains.yaml` is supposed to tell you about all of them, and west is still able to verify that each one is a valid Zephyr build.

Correct, there are multiple build systems, just like Zephyr uses external project when including TF-M in a build.
And you can even enter TF-M build folder to build TF-M even if the outer Zephyr CMake fails, but even if the TF-M may be able to succeed then I don't see the possibility to specifically build that using `west` as a use-case to cover.

Currently when using sysbuild, then each image is very independent, but it may well be that in future the output of one image is consumed by another image.
Therefore i'm concerned with giving the impression that `west build` is able to build some images if the high-level sysbuild has failed. Users with such requests must resort to their native build command (ninja / make), just like they must do in the TF-M case presented above.